diff --git a/scripts/mpc_node.py b/scripts/mpc_node.py
index 1cd2434..bd64068 100644
--- a/scripts/mpc_node.py
+++ b/scripts/mpc_node.py
@@ -17,13 +17,13 @@ from sensor_msgs.msg import LaserScan
 from visualization_msgs.msg import Marker, MarkerArray
 
 from utils import nearest_point
-
+from rclpy.qos import QoSProfile
 
 @dataclass
 class mpc_config:
     NXK: int = 4  # length of kinematic state vector: z = [x, y, v, yaw]
     NU: int = 2  # length of input vector: u = [steering speed, acceleration]
-    TK: int = 8  # finite time horizon length - kinematic
+    TK: int = 6  # finite time horizon length - kinematic
 
     # ---------------------------------------------------
     # TODO: you may need to tune the following matrices
@@ -36,11 +36,11 @@ class mpc_config:
         default_factory=lambda: np.diag([0.01, 100.0])
     )  # input difference cost matrix, penalty for change of inputs - [accel, steering_speed]
     Qk: list = field(
-        default_factory=lambda: np.diag([13.5, 13.5, 5.5, 13.0])  # levine sim
+        default_factory=lambda: np.diag([100.0, 100.0, 0.0, 0.0])  # levine sim
         # default_factory=lambda: np.diag([50., 50., 5.5, 13.0])
     )  # state error cost matrix, for the the next (T) prediction time steps [x, y, delta, v, yaw, yaw-rate, beta]
     Qfk: list = field(
-        default_factory=lambda: np.diag([13.5, 13.5, 5.5, 13.0])  # levine sim
+        default_factory=lambda: np.diag([100.0, 100.0, 0.0, 0.0])  # levine sim
         # default_factory=lambda: np.diag([50., 50., 5.5, 13.0])
     )  # final state error matrix, penalty  for the final state constraints: [x, y, delta, v, yaw, yaw-rate, beta]
     # ---------------------------------------------------
@@ -80,19 +80,19 @@ class MPC(Node):
         super().__init__('mpc_node')
         # use the MPC as a tracker (similar to pure pursuit)
         self.is_real = False
-        self.map_name = 'levine_2nd'
+        self.map_name = 'levine_full'
 
         # create ROS subscribers and publishers
-        pose_topic = "/pf/viz/inferred_pose" if self.is_real else "/ego_racecar/odom"
-        drive_topic = "/drive"
+        pose_topic = "/pf/viz/inferred_pose" # if self.is_real else "/ego_racecar/odom"
+        drive_topic = "/sim/drive"
         vis_ref_traj_topic = "/ref_traj_marker"
         vis_waypoints_topic = "/waypoints_marker"
         vis_pred_path_topic = "/pred_path_marker"
 
-        self.pose_sub = self.create_subscription(PoseStamped if self.is_real else Odometry, pose_topic, self.pose_callback, 1)
+        self.pose_sub = self.create_subscription(PoseStamped if not self.is_real else Odometry, pose_topic, self.pose_callback, 1)
         self.pose_sub  # prevent unused variable warning
 
-        self.drive_pub = self.create_publisher(AckermannDriveStamped, drive_topic, 1)
+        self.drive_pub = self.create_publisher(AckermannDriveStamped, drive_topic, QoSProfile(depth=1, reliability=1))
         self.drive_msg = AckermannDriveStamped()
 
         self.vis_waypoints_pub = self.create_publisher(Marker, vis_waypoints_topic, 1)
@@ -102,8 +102,8 @@ class MPC(Node):
         self.vis_pred_path_pub = self.create_publisher(Marker, vis_pred_path_topic, 1)
         self.vis_pred_path_msg = Marker()
 
-        map_path = os.path.abspath(os.path.join('src', 'csv_data'))
-        self.waypoints = np.loadtxt(map_path + '/' + self.map_name + '.csv', delimiter=';', skiprows=0)  # csv data
+        map_path = os.path.abspath(os.path.join('src', 'mpc'))
+        self.waypoints = np.loadtxt(map_path + '/racelines/' + self.map_name + '.csv', delimiter=';', skiprows=0)  # csv data
         if self.map_name == 'levine_2nd':
             self.waypoints[:, 3] += math.pi / 2
         self.visualize_waypoints_in_rviz()
@@ -163,7 +163,7 @@ class MPC(Node):
     # toolkits
     def update_rotation_matrix(self, pose_msg):
         # get rotation matrix from the car frame to the world frame
-        curr_orien = pose_msg.pose.orientation if self.is_real else pose_msg.pose.pose.orientation
+        curr_orien = pose_msg.pose.orientation if not self.is_real else pose_msg.pose.pose.orientation
         quat = [curr_orien.x, curr_orien.y, curr_orien.z, curr_orien.w]
         self.rot_mat = (transform.Rotation.from_quat(quat)).as_matrix()
         # print("rotation matrix = {}".format(self.rot_mat))
@@ -173,11 +173,11 @@ class MPC(Node):
         written by Derek, not from the template, != update state
         """
         vehicle_state = State()
-        vehicle_state.x = pose_msg.pose.position.x if self.is_real else pose_msg.pose.pose.position.x
-        vehicle_state.y = pose_msg.pose.position.y if self.is_real else pose_msg.pose.pose.position.y
+        vehicle_state.x = pose_msg.pose.position.x if not self.is_real else pose_msg.pose.pose.position.x
+        vehicle_state.y = pose_msg.pose.position.y if not self.is_real else pose_msg.pose.pose.position.y
         vehicle_state.v = self.drive_msg.drive.speed
 
-        curr_orien = pose_msg.pose.orientation if self.is_real else pose_msg.pose.pose.orientation
+        curr_orien = pose_msg.pose.orientation if not self.is_real else pose_msg.pose.pose.orientation
         q = [curr_orien.x, curr_orien.y, curr_orien.z, curr_orien.w]
         vehicle_state.yaw = math.atan2(2 * (q[3] * q[2] + q[0] * q[1]), 1 - 2 * (q[1] ** 2 + q[2] ** 2))
         # https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Quaternion_%E2%86%92_Euler_angles_(z-y%E2%80%B2-x%E2%80%B3_intrinsic)
